src/devices/machine/storyware.cpp

```diff
@@ -0,0 +1,190 @@
+{
+}
+
+void storyware_device::update_effective_page(uint32_t effective_page)
+{
+	m_effective_page = effective_page;
+}
+
+void storyware_device::update_pen_x(uint16_t x)
+{
+	m_pen_x = x;
+}
+
+void storyware_device::update_pen_y(uint16_t y)
+{
+	m_pen_y = y;
+}
+
+void storyware_device::create_pages()
+{
+	file_enumerator path(m_booklet_path);
+	std::vector<std::string> names;
+	for (osd::directory::entry const *dir = path.next(); dir; dir = path.next()) {
+		std::string name(dir->name);
+		if (core_filename_ends_with(name, ".png")) {
+			names.push_back(name);
+		}
+	}
+
+	// Since directory entries aren't iterated in order,
+	// ensure their names are sorted before pages are allocated.
+	std::sort(names.begin(), names.end());
+
+	m_page_count = 0;
+	for (std::string name : names) {
+		emu_file file(m_booklet_path, OPEN_FLAG_READ);
+		if (!file.open(name)) {
+			osd_printf_verbose("Loading page '%s'...\n", name);
+
+			bitmap_argb32 page_bitmap;
+			render_load_png(page_bitmap, file);
+
+			m_page_bitmaps[m_page_count].allocate(m_screen_w / 2, m_screen_h);
+			render_resample_argb_bitmap_hq(
+				m_page_bitmaps[m_page_count],
+				page_bitmap,
+				render_color{ 1.0F, 1.0F, 1.0F, 1.0F },
+				false);
+
+			file.close();
+
+			m_page_count++;
+		}
+	}
+}
+
+uint32_t storyware_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	bitmap.fill(rgb_t::white(), cliprect);
+
+	bitmap_argb32 *left_bitmap = nullptr;
+	bitmap_argb32 *right_bitmap = nullptr;
+	int8_t page_i = 2 * m_effective_page - 1;
+	if (page_i > m_page_count) {
+		page_i = m_page_count - 1;
+	}
+	if (page_i == -1) {
+		// If the storyware is closed, only the front cover is rendered.
+		if (m_page_bitmaps[0].valid()) {
+			right_bitmap = &m_page_bitmaps[0];
+		}
+	} else {
+		// If the storyware is open, show the back of the previous page and the front of the next page.
+		// The last effective page can vary across games, but in those cases we always show the back of the last page.
+		if (m_page_bitmaps[page_i].valid()) {
+			left_bitmap = &m_page_bitmaps[page_i];
+		}
+		if (m_page_bitmaps[page_i + 1].valid()) {
+			right_bitmap = &m_page_bitmaps[page_i + 1];
+		}
+	}
+	for (int y = cliprect.min_y; y <= cliprect.max_y; y++) {
+		if (left_bitmap != nullptr) {
+			for (int x = cliprect.min_x; x < (cliprect.max_x + 1) / 2; x++) {
+				bitmap.pix(y, x) = (*left_bitmap).pix(y, x);
+			}
+		}
+		if (right_bitmap != nullptr) {
+			for (int x = (cliprect.max_x + 1) / 2; x <= cliprect.max_x; x++) {
+				bitmap.pix(y, x) = (*right_bitmap).pix(y - 1, x);
+			}
+		}
+	}
+
+	draw_cursor(bitmap, cliprect, 0, 0xff/2);
+
+	return 0;
+}
+
+static const uint8_t cursor_bitmap[] = {
+	0, 0, 0, 1, 1, 1, 0, 0, 0,
+	0, 0, 2, 2, 2, 2, 2, 0, 0,
+	0, 2, 2, 2, 2, 2, 2, 2, 0,
+	1, 2, 2, 2, 2, 2, 2, 2, 1,
+	1, 2, 2, 2, 2, 2, 2, 2, 1,
+	1, 2, 2, 2, 2, 2, 2, 2, 1,
+	0, 2, 2, 2, 2, 2, 2, 2, 0,
+	0, 0, 2, 2, 2, 2, 2, 0, 0,
+	0, 0, 0, 1, 1, 1, 0, 0, 0,
+};
```

Comment:
From what I understood, it would be similar to `render_crosshair::create_bitmap`, but instead of reading from a file, just use the bitmap passed as an argument. I'm fine with either option, I'd even prefer reading from the file, since I can just specify it for the driver by putting it under `crosshairpath/beena/cross1.png`.

However, to use the crosshair, I would also need to adjust the position so that it maps to movement done only in the first screen. But now that I looked at this code again, I noticed that `ioport_field::crosshair_read` can overwrite the position with a `PORT_CROSSHAIR_MAPPER`. I'll have to give it a try to see if it fits.

Reply:
Managed to get it working and remove the custom cursor, now it's using the lightgun crosshair.

Apparently this mapper was unused, I had to update the arguments since they no longer matched any of the `device_delegate` constructors.